source/volume_of_fluid/reconstruct.cc

```diff
@@ -0,0 +1,488 @@
+    std::vector<typename DoFHandler<dim>::active_cell_iterator> neighbor_cells(n_local);
+
+    const unsigned int n_sums = 3;
+    std::vector<double> strip_sums (dim * n_sums);
+
+    const unsigned int n_normals = 6+1;
+    std::vector<Tensor<1, dim, double>> normals (n_normals);
+    std::vector<double> d_vals (n_normals);
+    std::vector<double> errs (n_normals);
+
+    // Variables to do volume calculations
+
+    QGauss<dim> quadrature(max_degree);
+
+    std::vector<double> xFEM_values(quadrature.size());
+
+    const FiniteElement<dim> &system_fe = this->get_fe();
+
+    FEValues<dim> fevalues(this->get_mapping(), system_fe, quadrature,
+                           update_JxW_values);
+
+    // Normal holding vars
+    Point<dim> uReCen;
+    Tensor<1, dim, double> normal;
+    double d;
+
+    for (unsigned int i=0; i<dim; ++i)
+      uReCen[i] = 0.5;
+
+    std::vector<types::global_dof_index> cell_dof_indicies (system_fe.dofs_per_cell);
+    std::vector<types::global_dof_index> local_dof_indicies (system_fe.dofs_per_cell);
+
+    const FEVariable<dim> &volume_of_fluid_var = field.volume_fraction;
+    const unsigned int volume_of_fluid_c_index = volume_of_fluid_var.first_component_index;
+    const unsigned int volume_of_fluid_ind
+      = this->get_fe().component_to_system_index(volume_of_fluid_c_index, 0);
+
+    const FEVariable<dim> &volume_of_fluidN_var = field.reconstruction;
+    const unsigned int volume_of_fluidN_c_index = volume_of_fluidN_var.first_component_index;
+    const unsigned int volume_of_fluidN_blockidx = volume_of_fluidN_var.block_index;
+
+    const FEVariable<dim> &volume_of_fluidLS_var = field.level_set;
+    const unsigned int volume_of_fluidLS_c_index = volume_of_fluidLS_var.first_component_index;
+    const unsigned int n_volume_of_fluidLS_dofs = volume_of_fluidLS_var.fe->dofs_per_cell;
+    const unsigned int volume_of_fluidLS_blockidx = volume_of_fluidLS_var.block_index;
+
+    //Iterate over cells
+    for (auto cell : this->get_dof_handler().active_cell_iterators ())
+      {
+        if (!cell->is_locally_owned ())
+          continue;
+
+        double cell_volume_of_fluid;
+
+        // Obtain data for this cell and neighbors
+        cell->get_dof_indices (local_dof_indicies);
+        cell_volume_of_fluid = solution(local_dof_indicies[volume_of_fluid_ind]);
+
+        normal[0] = 0.0;
+        normal[1] = 0.0;
+        d = -1.0;
+
+        if (cell_volume_of_fluid > 1.0 - volume_fraction_threshold)
+          {
+            d = 1.0;
+            initial_solution(local_dof_indicies[volume_of_fluid_ind]) = 1.0;
+          }
+        else if (cell_volume_of_fluid < volume_fraction_threshold)
+          {
+            d = -1.0;
+            initial_solution(local_dof_indicies[volume_of_fluid_ind]) = 0.0;
+          }
+        else
+          {
+            initial_solution(local_dof_indicies[volume_of_fluid_ind]) = cell_volume_of_fluid;
+
+            //Identify best normal
```

Comment:
Please unify spaces at the beginning of comments, I prefer a space, but either is fine as long as it is uniform.